File: hw2/crouts.m
function [L, U] = crouts(A)
    [m, n] = size(A);
    
    if m == 1
        L = A;
        U = 1;
        return;
    end
    
    [L_sub, U_sub] = crouts(A(2:m, 2:n));
    
    L = zeros(m, m);
    U = eye(m, n);
    
    % First column of L and first row of U
    L(:,1) = A(:,1);
    U(1,2:n) = A(1,2:n) / L(1,1);
    
    % Fill submatrices
    L(2:m, 2:m) = L_sub;
    U(2:m, 2:n) = U_sub;
    
    % Update L's first column properly
    for i = 2:m
        for j = 1:i-1
            L(i,1) = L(i,1) - L(i,j)*U(j,1);
        end
    end
    
    % Update last row of U
    U(m,2:n) = (A(m,2:n) - L(m,1:m-1)*U(1:m-1,2:n));
end

File: hw2/crouts.py
import numpy as np


def crouts(A):
    if not isinstance(A, np.ndarray):
        A = np.array(A, dtype=float)
    n = len(A)

    # Base case
    if n == 1:
        return np.array([[A[0, 0]]]), np.array([[1]])

    # Initialize L and U
    L = np.zeros((n, n))
    U = np.eye(n)

    # First column of L is directly from A
    L[:, 0] = A[:, 0]

    # First row of U (except U[0,0]=1)
    U[0, 1:] = A[0, 1:] / L[0, 0]

    A_reduced = A[1:, 1:] - np.outer(L[1:, 0], U[0, 1:])
    L_sub, U_sub = crouts(A_reduced)

    # Fill in submatrices
    L[1:, 1:] = L_sub
    U[1:, 1:] = U_sub

    return L, U
